Fetch dataset

allthatilk opened the pull request.

dalepotter committed:
Add initial requirements files

These are set-up as a stub and only test packages are required
at this time.

dalepotter committed:
Make iati.fetch into a package

dalepotter committed:
Remove unnecessary comment

dalepotter committed:
Add initial stub code to fetch a dataset metadata by a dataset_id

Tests do not pass, yet!

dalepotter committed:
Add version support and install info to README

Commit:
Fix linting errors

allthatilk committed:
Fix iati.core git repo dependency in setup

allthatilk committed:
Exclude virtualenv from pytest dir

allthatilk committed:
Functionality for fetching valid dataset

allthatilk committed:
Refactoring and add mock of registry metadata

dalepotter committed:
Beautify python dict

dalepotter committed:
Blacklist -> whitelist test directories

dalepotter committed:
Fix filename for dev requirements

dalepotter committed:
Setup Travis CI

dalepotter committed:
Use requests_mock decorator, rather than with statement

This is an attempt to make tests using mocking more DRY in advance of
adding additional, similar tests.

dalepotter committed:
Execute py.test in Travis CI

dalepotter committed:
Amend Registry API endpoint to a constant

dalepotter committed:
Improve docstrings

Adds 'Args', 'Raises', 'Returns' and 'Todo' elements to function and method docstrings.

allthatilk committed:
Merge pull request #1 from IATI/extract-api-endpoint-to-constant

Amend Registry API endpoint to a constant

allthatilk committed:
Merge branch 'fetch-dataset' into improve-docstrings

allthatilk committed:
Merge pull request #2 from IATI/improve-docstrings

Improve docstrings

hayfield commented on .gitignore:
```diff
@@ -84,6 +84,7 @@ celerybeat-schedule
 
 # virtualenv
 .venv
+pyenv/
```
Since this is custom rather than part of https://github.com/github/gitignore/blob/master/Python.gitignore, it should be at the top of the file

hayfield commented on iati/fetch/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    """
+    req = requests.get(REGISTRY_API_METADATA_BY_DATASET_ID.format(dataset_id))
+
+    if not is_response_okay(req):
```
The standard case (all is well) should be first

hayfield replied:
The same goes for other `if` statements

hayfield commented on iati/fetch/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    Returns:
+        str: The data contained at the dataset_url.
+    """
+    dataset = requests.get(dataset_url)
```
The `dataset` variable contains a response object rather than a Dataset, so should probably indicate this in the name somehow.

hayfield commented on iati/fetch/tests/test_dataset.py:
```diff
@@ -0,0 +1,35 @@
+        """Given a dataset_id, metadata for this dataset will be populated.
+
+        Todo:
+          Remove use of live dataset_id.
```
Level of indentation lower than elsewhere in the file

hayfield commented on requirements-dev.txt:
```diff
@@ -0,0 +1,13 @@
+# testing tools
+pytest==3.0.5
+pytest-cov==2.4.0
+requests_mock
```
Requirements should have versions against them to help improve deployment repeatability.

hayfield commented on requirements-dev.txt:
```diff
@@ -0,0 +1,13 @@
+#
```
Suggest adding requires.io filters to pin requirements at the current integer (automation and backwards compatibility!)
https://requires.io/features/

hayfield commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+#
+# Install this package
+#
+-e git://github.com/IATI/iati.core.git#egg=iati.core
```
Oooo, an actual use of eggs, whatever they are!

hayfield commented on iati/fetch/__init__.py:
```diff
@@ -0,0 +1,84 @@
+REGISTRY_API_METADATA_BY_DATASET_ID = "https://iatiregistry.org/api/3/action/package_show?id={0}"
+
+
+class Dataset(object):
```
May want to inherit from `iati.core.dataset.Dataset`

hayfield commented on iati/fetch/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    req = requests.get(REGISTRY_API_METADATA_BY_DATASET_ID.format(dataset_id))
+
+    if not is_response_okay(req):
+        raise Exception()
```
Could raise a more specific Exception

hayfield commented on iati/fetch/__init__.py:
```diff
@@ -0,0 +1,84 @@
+
+
+def get_metadata(dataset_id):
+    """Get registry metadata for given dataset ID.
```
Need to determine preferred capitalisation of terms such as `Dataset`, `Schema`, `Codelist`, etc that are used within `iati.*`

hayfield commented on iati/fetch/__init__.py:
```diff
@@ -0,0 +1,84 @@
+        bool: True is the HTTP status code is 200, or False otherwise.
+
+    Todo:
+        Delete this function and use builtin requests.status.ok instead.
```
Should be `request.ok` assuming the mocked version works as per the full version

hayfield reviewed (commented):
Some super nitpicky comments and suggestions!

dalepotter committed:
Add stubs for further tests

dalepotter committed:
Abstract Registry API endpoint

This means the appropriate Registry API endpoint can be passed in as a function
parameter.

allthatilk committed:
Merge pull request #4 from IATI/abstract-url

Abstract Registry API endpoint

allthatilk committed:
Refactor mock using registry URL injection

allthatilk committed:
Refactor mock using registry URL injection

dalepotter committed:
Add test for iati.fetch.get_dataset()

Tests response from iati.fetch.get_dataset() for expected number of
lines and expected strings contained within returned data.

dalepotter committed:
Move decorator to class level

dalepotter committed:
Make mock dataset URL clearer

allthatilk committed:
Invalid ID raises exception as expected.

allthatilk committed:
Removing unneccesary incomplete method in __init__

allthatilk committed:
Add todo to docstring

dalepotter committed:
Merge pull request #6 from IATI/invalid-id

Invalid ID

dalepotter committed:
Fix merge conflicts and move decorator

As part of the merge from fetch-dataset into this branch, the
@requests_mock.mock decorator has been moved back to method
level.

dalepotter committed:
Improve how dataset encoding is handled

This is necessary as str.encode(utf-8) seems to return a bytes object,
which seems counterintuitive.

allthatilk committed:
Merge pull request #7 from IATI/add-test-getdataset

Add test for iati.fetch.get_dataset and improve encoding handling

allthatilk committed:
Uses responses modulefor mocking requests

dalepotter committed:
Swap requests_mock to responses for further test

dalepotter committed:
Add space when handling Exception

dalepotter committed:
Improve processing of encoded data

Makes the if/else statement more robust and attempts return data in a string format.

dalepotter committed:
Swap final test from requests_mock to responses

Also removes requests_mock from development dependencies file.

dalepotter committed:
Swap use to bytes() to a crude string replace

The bytes() function requires an `encoding` param in Python3, but
this is not allowed in Python2. Therefore, make a crude string
replace until IATI/iati.core#24 is resolved.

dalepotter committed:
Remove resolved Todo comment from docstring

allthatilk committed:
Merge pull request #8 from IATI/use-responses

Uses responses modulefor mocking requests

allthatilk committed:
Add test for bad dataset url

allthatilk committed:
Add decorator to call set_dataset upon instantiation

allthatilk committed:
Remove commented out line

allthatilk committed:
Add mock registry url constant for maintainability

dalepotter committed:
Add function to return all IATI publishers

allthatilk committed:
Change mock URL in line with rfc2606

dalepotter committed:
Fix failing Travis test - Exception test varies between versions

dalepotter committed:
Restructure asserts given python version differences

Python v3.4 returns 'ValueError: Expecting value' whereas v3.5 returns 'JSONDecodeError: Expecting value'

allthatilk committed:
Replace decorator with better code

hayfield committed:
Make set_dataset() set something

This was easier than thinking of a name for the other implementation.

allthatilk committed:
Merge pull request #10 from IATI/add-decorator

Add decorator

allthatilk committed:
Merge pull request #11 from IATI/mock-url-constant

Mock url constant

allthatilk committed:
Merge pull request #9 from IATI/bad-url

Bad url

dalepotter committed:
Merge branch 'fetch-dataset' into get-publishers

dalepotter committed:
Swap hardcoded URL to use MOCK_REGISTRY_URL constant

allthatilk committed:
Merge pull request #12 from IATI/get-publishers

Add iati.fetch.get_publishers
